### ISSUE TITLE
Combobox with many values has buggy selection behaviour on ipad

### ISSUE THREAD

**Issue by nickknw** [jquense/react-widgets#113]
The <input> element is normally (on desktop browsers) given focus when the combobox's dropdown list is *expanded*. On the iPad it doesn't receive focus until one of the items from the dropdown list is *selected*, which pops up the on screen keyboard and shifts the browser page so that the wrong item ends up getting selected when you remove your finger. 

This can be worked around by focusing the input box first and then expanding the list.

EDIT: It actually focuses the input box when you tap *anywhere* after expanding the list, does not have to be a list item. So another workaround would be to expand the list, tap elsewhere, then select the desired item.

One fix would be to focus the input box as soon as the dropdown is expanded, to match desktop behaviour. If a user is interested in only the dropdown part of the combobox this means it will be a bit more difficult to use (with the keyboard taking up half the screen when it is not necessarily needed). This is the approach I think makes most sense though.

The other fix would be to avoid focusing the input box when a list item is selected. This is a change to typical combobox behaviour though and likely to cause more focus related problems. Comboboxes are used where you need to be able to type something, don't want to make it harder to do that.

Tested with iPad IOS 8.1

**Comment by jquense** [comment]
Thanks for this,

I don't have an IOS devices so its hard for me to try and fix this. I am not sure why focus wouldn't be captured, Either nothing inside the combobox wrapper receives focus, or something is stealing focus outside the wrapper. Both seem unlikely. It sounds like mobile safari is doing something really off spec...

which is all to say that it is trying to do the right thing, but there seems to be an IOS specific issue that I don't know about and will have a hard time tracking down. I would appreciate any input/help if you have any though!

**Comment by nickknw** [comment]
Yeah I'm not sure what's going on either yet, but I'm looking into it. Will let you know what I find out!

**Comment by nickknw** [comment]
Yep, looks like mobile safari has some special restrictions around focusing input elements. ["iOS will only allow focus to be triggered on other elements, from within a function, if the first function in the call stack was triggered by a non-programmatic event"](http://stackoverflow.com/q/6287478/224354). Calling setTimeout starts a new call stack and so calling focus() from within there is ignored. ([Another SO post on same issue](http://stackoverflow.com/a/16601288/224354))

Apparently there was a change in this behaviour in iOS 5 that should let handlers triggered from synthesized click events trigger a focus in as well, although I tried that a few ways in a jsfiddle and couldn't get it to work. The post might be talking about [this setting that was introduced](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIWebView_Class/index.html#//apple_ref/occ/instp/UIWebView/keyboardDisplayRequiresUserAction), which technically does let you do that, but doesn't really help us.

What did work for me was triggering an additional focus before or after the setTimeout call. Since it was still in the user-initiated call stack the focus worked just fine.

I see in Combobox that the focus is triggered inside a setTimeout. One solution might be to trigger it twice - once before (for iOS) and once during (for everything else). In most cases that should work fine, although I could see it causing other edge-casey focus issues. If you have a method of browser detection available inside the library maybe it could be restricted to iOS only? Not sure what the right solution is here. Any thoughts?

**Comment by jquense** [comment]
Hey thanks so much for looking into this, it is extremely helpful. I think we can get away with calling the focus outside of the timeout, which is really there to make sure the blurs can be Canceled by the next focus if the focus is inside the widget container. Let me go with that and I'll see if I can just do it for IOS

**Comment by jquense** [comment]
Hey @nickknw I've made a few adjusts that should push the focus trigger into the right place, can you see if it resolves the issue for you? You may need to build the compiled files yourself tho `gulp lib` should do it

**Comment by nickknw** [comment]
Hey sorry for the delay, recovering from quite a flu. 

I have been unable to build the compiled files locally, unfortunately! Whenever I try to install your forked version of rf-release I get this error:

```
$ npm install
npm WARN package.json node-libs-browser@0.5.0 No repository field.
npm ERR! Darwin 14.3.0
npm ERR! argv "node" "/usr/local/bin/npm" "install"
npm ERR! node v0.12.0
npm ERR! npm  v2.7.5
npm ERR! path /Users/nick.knowlson/workspace/react-widgets/node_modules/rf-release/bin/rf-release
npm ERR! code ENOENT
npm ERR! errno -2

npm ERR! enoent ENOENT, chmod '/Users/nick.knowlson/workspace/react-widgets/node_modules/rf-release/bin/rf-release'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/nick.knowlson/workspace/react-widgets/npm-debug.log
```
npm-debug.log is about 6000 lines but I could send that over as well if you like.

Have tried updating npm, deleting node_modules and clearing npm cache. Not sure why it is failing. Does this look familiar to you / can you shed any light on it?

**Comment by jquense** [comment]
Oof I'm not sure. npm hates installing forks sometimes. let me just rebuild the stuff and push it so you don't need to

**Comment by nickknw** [comment]
Awesome, thank you!

**Comment by nickknw** [comment]
Can confirm these changes fix the issue. Thanks!

**Comment by jquense** [comment]
Woohoo! thanks a ton for the debugging and tracking this down. It wouldn't have been fixed otherwise :)

**Comment by LeoIannacone** [comment]
Hi @jquense ..
Can you please make a release with this fix?

**Comment by jquense** [comment]
Sorry folks, I'll get a release out soon,  next week some time

**Comment by LeoIannacone** [comment]
Hi @jquense ... sorry for boring... Any news?

**Comment by jquense** [comment]
@LeoIannacone fixed in 2.6.1

**Comment by LeoIannacone** [comment]
thanks !
